### ISSUE TITLE
Inner classes in stubs files

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#568]
Is it possible to annotate qualified inner class types in stub files?

The manual [mentions](http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#stub-limitations) that the stub parser can't handle annotations inside qualified names, and that `$`-qualified inner class names can be used as a workaround. I can't get it to work for parameter types.

Repro:

Lib.java

```java
import org.checkerframework.checker.nullness.qual.Nullable;
class Lib {
  static class Inner {}
  static void f(Lib.Inner e) {}
}
```

Test.java

```java
class Test {
  {
    Lib.f(null);
  }
}
```

stubs/lib.astub

```java
import org.checkerframework.checker.nullness.qual.Nullable;
class Lib {
  static void f(Lib.Inner e) {}
}
```

A few options I tried:

`static void f(Lib.@Nullable Inner e) {}`

Fails with:

```
  Underlying Exception: org.checkerframework.stubparser.ParseException: Encountered " "class" "class "" at line 3, column 1.
  Was expecting one of:
      <EOF>
      "import" ...
      "package" ...
      "@" ...
      "\u001a" ...
      "@" ...
      "@" ...
      "@" ...
      "@" ...
      ; Stack trace: org.checkerframework.stubparser.ASTParser.generateParseException(Unknown Source
```

`static void f(@Nullable Lib.Inner e) {}`

Fails with: `error: [argument.type.incompatible] incompatible types in argument.`

`static void f(@Nullable Lib$Inner e) {}`

Fails with: `error: [argument.type.incompatible] incompatible types in argument.`

**Comment by smillst** [comment]
Thanks for reporting this.  The following seems to work:
```
static void f(@Nullable Inner e) {}
```
I'm not sure if this is a "bug" or an expected limitation.  If it's a limitation, we need to document it.  

The quote below is from [the stub file section of the manual](http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#stub-limitations)
>The stub file reader does not handle nested classes. To work around this, it permits a top-level class to be written with a $ in its name, and applies the annotations to the appropriate nested class.

This limitation is about annotating an inner class or methods in an inner class in stubfiles.  For example:
````
@SomeAnnotation
class Lib$Inner{
}
````
We should clarify this.

**Comment by cushon** [comment]
Thanks, clarifying the docs would be great.

Importing the inner type directly should work for the example that inspired this bug, but occasionally it's necessary to disambiguate two inner classes with the same simple name. It'd be nice if that could be expressed with stubs.

**Comment by smillst** [comment]
I dug into this a bit further and found the following:
If the class `Lib` is in package `mypackage`, then you can annotation the `f`'s parameter in a stub file like so:
````
static void f(@Nullable mypackage.Lib.Inner e) {}
````
(I'll add this to the manual.)

However, the stub parser does not differentiate between the two `g` methods in `MyClass` shown below.  Do you anticipate needing to annotate such methods in a stub file? 
````
public class MyClass {
  public static class Inner {}
  public static void g(Lib.Inner e) {}
  public static void g(MyClass.Inner e) {}
}
````

**Comment by smillst** [comment]
(Pull request #570 clarifies the manual)

**Comment by cushon** [comment]
Thanks, using fully qualified names to disambiguate inner classes should be fine.

If the stub parser does not differentiate between those versions of `g`, can it report an error?

**Comment by smillst** [comment]
>If the stub parser does not differentiate between those versions of g, can it report an error?

tl;dr Unfortunately no, the correct fix is to change the stub parser to use the latest version of [JavaParser](http://javaparser.github.io/javaparser/).

It's actually the version of [JavaParser](http://javaparser.github.io/javaparser/) that is used by the stub parser that cannot differentiate between the two.  The version used is 5+ years old and only supports Java 5 syntax. (See the [README](https://github.com/typetools/checker-framework/blob/master/stubparser/README) in the stubparser package. The correct fix for all of the stub file limitations is to use the latest version of JavaParser; we just haven't taken the time to do so.)

Because of this I don't see an easy way to issue an error if either `g` method is used in a stub file.  However, if both methods were used in a stub file 
````
public class MyClass {
  public static void g(mypackage.Lib.Inner e) {}
  public static void g(mypackage.MyClass.Inner e) {}
}
````

you'll get this error:
````
error: InvocationTargetException when invoking constructor for class org.checkerframework.checker.nullness.KeyForVisitor; Underlying cause: InvocationTargetException when invoking constructor for class org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory; Underlying cause: StubParser: key is already in map:
    h(mypackage.Lib.Inner) => static void g(@Nullable mypackage.Lib.Inner e) {
  }
  while adding:
    h(mypackage.Lib.Inner) => static void g(@Nullable mypackage.MyClass.Inner e) {
  }; invoke the compiler with -AprintErrorStack to see the stack trace.
1 error
```` 
The key in this map is the element for the method 'g'.  Apparently both 'g' methods have the same element supplied by (this old versions of) JavaParser.

If only one 'g' method is in the stub file, then it's annotations are applied to the `g` that comes first in the class file. I don't see a way to detect this case.

**Comment by smillst** [comment]
The documentation updates were merged when pull request #570 was merged (a54416c).  Another issue (#571) has been opened to address the underlying issue.
